### ISSUE TITLE
Output diff not generated in fresh build

### ISSUE THREAD

**Issue by opatry** [nanoc/nanoc#1527]
### Steps to reproduce

1. Clean your output (`rm -rf output/ tmp/`)
2. Compile your site with diff support (`bundle exec nanoc compile --diff`)

### Expected behavior

`output.diff` file generated with everything considered "added".

### Actual behavior

No `output.diff` file generated.

### Details

I'd like to condition the site deployment to the presence of `output.diff`.
On my CI, I have a cache of the `output/` restored on new build (using GitHub Actions `actions/cache@v2` with `path: output`).
Using nanoc's output diff mechanism, I can determine if the new build changed output or not.
This won't work when no cache is present (evicted, first time, …).

Wouldn't be better to generate the diff file in such case?

It's a suggestion I share, I can define a dedicated solution for "fresh build" situation in my case but I guess this would make sense.

**Comment by ddfreyne** [comment]
Good point — and I think the change should be fairly small. See `item_rep_writer.rb`:

```ruby
        # Notify ready for diff generation
        if !is_created && is_modified && !content.binary?
          Nanoc::Core::NotificationCenter.post(
            :rep_ready_for_diff, raw_path, File.read(raw_path, encoding: 'UTF-8'), content.string
          )
        end
```

The `rep_ready_for_diff` notification kicks off diff generation (see `diff_generator.rb`).

It might be as simple as changing the condition from `!is_created && is_modified` to `is_created || is_modified`, but I’ve not tested it out.

Binary items will likely still be a problem for you, though, as Nanoc doesn’t generate diffs for binary files.

**Comment by opatry** [comment]
In fact, it's worse than that, the output.diff is only generated when there is an `update`, not when `delete` and/or `create`.

So, alternative scenario:

1. Compile the site in a given output state A
2. Change a routing rule so that `output/file1.html` become `output/file2.html`
3. Compile with diff

Here, you'll have
```
      delete  /absolute/path/to/output/file1.html
      create  [0.03s]  output/file2.html
```

But no `output.diff` will be generated which is annoying and misleading here.

**Comment by opatry** [comment]
In the meantime, I'll use the following mechanism:
```
$ set -o pipefail; bundle exec nanoc compile |& tee output.txt
$ changes_count=$(grep -cw -e " create " -e " delete " -e " update " output.txt)
```

It works for fresh build (100% `create`), content change (`update`), file removal, addition or rename (`delete` & `create`).

**Comment by opatry** [comment]
I didn't pay attention to binary items (which I use a lot…), so you're right, that might be a problem.

I don't know if it's feasible nor desirable, but couldn't we compute the hash of the binary items and compare to include them in diff? Such (heavy?) computation could be done only when diff output is requested?

**Comment by opatry** [comment]
Given the discussion, if I change the approach and use `find output -type f -print0 | sort -z | xargs -0 shasum | shasum`, it seems I would get the expected result even with binary items.

**Comment by ddfreyne** [comment]
Labeling this as an enhancement rather than a bug, though the difference is a bit vague in this case.

**Comment by ddfreyne** [comment]
My thinking is that `output.diff` is only intended for human (rather than programmatic) usage, and that Nanoc would benefit from a `--json` flag for parseable output. If I’m not mistaken, that would fit your use case as well — correct?

**Comment by opatry** [comment]
The sole presence of `output.diff` is enough for my personal use case.
Having a JSON file I can query with tool like `jq` would be nice too.

That being said, I finally used `output/` hash comparison which is really simple and reliable for my use case (similar to GitHub `hashFiles()` but taking into account file names which is key for SSG deploy use case).

Still, having a consistent/predictable behavior of output diff is highly desirable for nanoc :)
